src/send_packets.c

```diff
@@ -560,35 +562,54 @@ send_packets(tcpreplay_t *ctx, pcap_t *pcap, int idx)
             update_flow_stats(ctx,
                     options->cache_packets ? sp : NULL, &pkthdr, pktdata, datalink);
 
+        if (ctx->first_time)
+        {
+            /* get time and timestamp of the first packet */
+            gettimeofday(&now, NULL);
+            memcpy(&first_pkt_ts, &pkthdr.ts, sizeof(struct timeval));
+        }
+
         /*
          * this accelerator improves performance by avoiding expensive
          * time stamps during periods where we have fallen behind in our
          * sending
          */
         if (skip_length && pktlen < skip_length) {
             skip_length -= pktlen;
-            now_is_now = false;
         } else if (ctx->skip_packets) {
             --ctx->skip_packets;
-            now_is_now = false;
         } else {
             /*
              * time stamping is expensive, but now is the
              * time to do it.
              */
+            dbgx(4, "This packet time: " TIMEVAL_FORMAT, pkthdr.ts.tv_sec, pkthdr.ts.tv_usec);
             skip_length = 0;
             ctx->skip_packets = 0;
-            now_is_now = true;
-            gettimeofday(&now, NULL);
 
             /*
-             * Only sleep if we're not in top speed mode (-t)
+             * Only sleep if we're not in top speed mode (-t) or late
              *
              * This also sets skip_length which will avoid timestamping for
              * a given number of packets.
              */
-            calc_sleep_time(ctx, &pkthdr.ts, &ctx->stats.last_time, pktlen, sp, packetnum,
-                    &ctx->stats.end_time, &start_us, &skip_length);
+            timersub(&pkthdr.ts, &first_pkt_ts, &pkt_ts_delta);
+            if (calc_sleep_time(ctx, &pkt_ts_delta, &ctx->stats.last_time, pktlen, sp, packetnum,
+                    &ctx->stats.end_time, &start_us, &skip_length))
+            {
```

Comment:
Slight request for style change to be consistent with project: brace moved up a line.